include/pstl_offload/internal/usm_memory_replacement.h

```diff
@@ -51,85 +52,148 @@ __get_offload_device_selector()
+
+        __scoped_lock(__scoped_lock&) = delete;
+        __scoped_lock&
+        operator=(__scoped_lock&) = delete;
+
+        ~__scoped_lock() { _M_mutex.__unlock(); }
+    };
+    friend class __scoped_lock;
+};
+
+static __spin_mutex __offload_policy_holder_mtx;
+
 class __offload_policy_holder_type
 {
-    using __set_active_device_func_type = void (*)(sycl::device*);
+    using __set_device_status_func_type = void (*)(bool);
 
   public:
-    // Since the global object of __offload_policy_holder_type is static but the constructor
+    // Since the global object of __offload_policy_holder_type is static but the template constructor
     // of the class is inline, we need to avoid calling static functions inside of the constructor
-    // and pass the pointer to exact function as an argument to guarantee that the correct __active_device
+    // and pass the pointer to exact function as an argument to guarantee that the correct offload device
     // would be stored in each translation unit
     template <typename _DeviceSelector>
     __offload_policy_holder_type(const _DeviceSelector& __device_selector,
-                                 __set_active_device_func_type __set_active_device_func)
-        : _M_set_active_device(__set_active_device_func)
+                                 __set_device_status_func_type __set_device_status_func)
+        : _M_set_device_status_func(__set_device_status_func)
     {
+        sycl::device _device;
+
         try
         {
-            _M_offload_device.emplace(__device_selector);
-            _M_set_active_device(&*_M_offload_device);
-            _M_offload_policy.emplace(*_M_offload_device);
+            _device = sycl::device(__device_selector);
         }
         catch (const sycl::exception& e)
         {
-            // __device_selector throws with e.code() == sycl::errc::runtime when device selection unable
+            // __device_selector call throws with e.code() == sycl::errc::runtime when device selection unable
             // to get offload device with required type. Do not pass an exception, as ctor is called for
             // a static object and the exception can't be processed.
-            // Remember the situation and re-throw exception when asked for the policy from user's code.
-            // Re-throw in every other case, as we don't know the reason.
-            if (e.code() != sycl::errc::runtime)
+            // Remember the situation as empty _M_device and re-throw exception when asked for
+            // the policy from user's code.
+            // Re-throw in every other case, as we don't know the reason of an exception.
+            if (e.code() == sycl::errc::runtime)
+            {
+                return;
+            }
+            else
+            {
                 throw;
+            }
         }
+
+        __spin_mutex::__scoped_lock __lock(__offload_policy_holder_mtx);
+
+        new (&_M_offload_device) __sycl_device_shared_ptr(_device);
```

Comment:
I believe there are more correctness issues, I am thinking about simpler implementation even if it means to increase the size of header.